From 8805dbc52d258ecf74b2359dcaff4c67cbaf161f Mon Sep 17 00:00:00 2001
From: Ville Tervo <ville.tervo@nokia.com>
Date: Wed, 2 Jun 2010 14:08:31 +0300
Subject: [PATCH] test valid clock_offset

---
 src/security.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/src/security.c b/src/security.c
index 652eea6..9dfb0fa 100644
--- a/src/security.c
+++ b/src/security.c
@@ -864,6 +864,7 @@ static inline void conn_complete(int dev, int dev_id, bdaddr_t *sba, void *ptr)
 	/* Request remote name */
 	memset(&cp_name, 0, sizeof(cp_name));
 	bacpy(&cp_name.bdaddr, &evt->bdaddr);
+	cp_name.clock_offset = htobs(0x0000);
 	cp_name.pscan_rep_mode = 0x02;
 
 	data = hci_req_data_new(dev_id, &evt->bdaddr, OGF_LINK_CTL,
-- 
1.7.0.1

