diff --git a/plugins/hciops.c b/plugins/hciops.c
index 9354bbe..596196e 100644
--- a/plugins/hciops.c
+++ b/plugins/hciops.c
@@ -3493,8 +3493,10 @@ static void bonding_connect_cb(GIOChannel *io, GError *err, gpointer user_data)
 		/* Wait proper error to be propagated by bonding complete */
 		return;
 
-	if (request_authentication(dev->id, &conn->bdaddr) < 0)
-		goto failed;
+	/* Removing the authentication request from userspace since it
+	   should be made by the kernel. The modification can't be
+	   pushed upstream (yet?) since without the corresponding
+	   changes in the kernel, legacy pairing won't work */
 
 	return;
 
