diff --git a/plugins/pnat.c b/plugins/pnat.c
index 6f02f51..2d73910 100644
--- a/plugins/pnat.c
+++ b/plugins/pnat.c
@@ -94,14 +94,15 @@ static void disconnect(struct dun_server *server)
 	if (!client->io)
 		return;
 
-	g_io_channel_unref(client->io);
-	client->io = NULL;
-
 	if (client->io_watch > 0) {
 		g_source_remove(client->io_watch);
 		client->io_watch = 0;
 	}
 
+	g_io_channel_shutdown(client->io, TRUE, NULL);
+	g_io_channel_unref(client->io);
+	client->io = NULL;
+
 	if (client->pnatd_pid > 0) {
 		kill(client->pnatd_pid, SIGTERM);
 		client->pnatd_pid = 0;
