diff --git a/src/event.c b/src/event.c
index 11f153e..764ebb5 100644
--- a/src/event.c
+++ b/src/event.c
@@ -790,8 +790,16 @@ void btd_event_conn_complete(bdaddr_t *local, uint8_t status, bdaddr_t *peer)
 	struct btd_adapter *adapter;
 	struct btd_device *device;
 	DBusConnection *conn = get_dbus_connection();
+	gboolean create;
 
-	if (!get_adapter_and_device(local, peer, &adapter, &device, TRUE))
+	DBG("status=%02x", status);
+
+	create = status ? FALSE : TRUE;
+
+	if (!get_adapter_and_device(local, peer, &adapter, &device, create))
+		return;
+
+	if (!device)
 		return;
 
 	if (status) {
